pinot-core/src/main/java/org/apache/pinot/core/operator/docidsets/AndDocIdSet.java

```diff
@@ -0,0 +1,156 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.docidsets;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.pinot.common.utils.Pairs.IntPair;
+import org.apache.pinot.core.common.BlockDocIdIterator;
+import org.apache.pinot.core.operator.dociditerators.AndDocIdIterator;
+import org.apache.pinot.core.operator.dociditerators.BitmapBasedDocIdIterator;
+import org.apache.pinot.core.operator.dociditerators.RangelessBitmapDocIdIterator;
+import org.apache.pinot.core.operator.dociditerators.ScanBasedDocIdIterator;
+import org.apache.pinot.core.operator.dociditerators.SortedDocIdIterator;
+import org.apache.pinot.core.util.SortedRangeIntersection;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+/**
+ * The FilterBlockDocIdSet to perform AND on all child FilterBlockDocIdSets.
+ * <p>The AndBlockDocIdSet will construct the BlockDocIdIterator based on the BlockDocIdIterators from the child
+ * FilterBlockDocIdSets:
+ * <ul>
+ *   <li>
+ *     When there are at least one index-base BlockDocIdIterators (SortedDocIdIterator or BitmapBasedDocIdIterator) and
+ *     at least one ScanBasedDocIdIterator, or more than one index-based BlockDocIdIterators, merge them and construct a
+ *     RangelessBitmapDocIdIterator from the merged document ids. If there is no remaining BlockDocIdIterators, directly
+ *     return the merged RangelessBitmapDocIdIterator; otherwise, construct and return an AndDocIdIterator with the
+ *     merged RangelessBitmapDocIdIterator and the remaining BlockDocIdIterators.
+ *   </li>
+ *   <li>
+ *     Otherwise, construct and return an AndDocIdIterator with all BlockDocIdIterators.
+ *   </li>
+ * </ul>
+ */
+public final class AndDocIdSet implements FilterBlockDocIdSet {
+  private final List<FilterBlockDocIdSet> _docIdSets;
+
+  public AndDocIdSet(List<FilterBlockDocIdSet> docIdSets) {
+    _docIdSets = docIdSets;
+  }
+
+  @Override
+  public BlockDocIdIterator iterator() {
+    int numDocIdSets = _docIdSets.size();
+    // NOTE: Keep the order of FilterBlockDocIdSets to preserve the order decided within FilterOperatorUtils.
+    // TODO: Consider deciding the order based on the stats of BlockDocIdIterators
+    BlockDocIdIterator[] allDocIdIterators = new BlockDocIdIterator[numDocIdSets];
+    List<SortedDocIdIterator> sortedDocIdIterators = new ArrayList<>();
```

Comment:
These 3 (sorted, inv index, and scan) are basically for left or right leaf operators of AND. The `remainingDocIdIterators` is for non-leaves (child AND/OR) right? 

On that note, for any sub-tree rooted at AND, there can be at-most one child with sorted iterator. Right?

Reply:
For the current supported iterators, yes `remainingDocIdIterators` can only be AND/OR.

There could be multiple sorted iterators if there are multiple predicates on the same sorted column or there are multiple sorted columns.